odoo/tools/template_inheritance.py

```diff
@@ -205,23 +205,94 @@ def extract(spec):
                     raise ValueError(_("Invalid mode attribute:") + " '%s'" % mode)
             elif pos == 'attributes':
                 for child in spec.getiterator('attribute'):
-                    attribute = child.get('name')
-                    value = child.text or ''
-                    if child.get('add') or child.get('remove'):
-                        assert not child.text
-                        separator = child.get('separator', ',')
-                        if separator == ' ':
-                            separator = None    # squash spaces
-                        to_add = (
-                            s for s in (s.strip() for s in child.get('add', '').split(separator))
-                            if s
+                    attribute = child.attrib.pop('name')
+                    value = None
+
+                    if child.get('add') or child.get('remove') or 'separator' in child.attrib:
+                        if child.text:
+                            raise ValueError(
+                                _("Forbidden text %s") %
+                                repr(child.text)
+                            )
+                        value = node.get(attribute, '')
+
+                        if attribute in ('invisible', 'column_invisible', 'required', 'readonly', 't-if', 't-elif') or attribute.startswith('decorator-'):
+                            # python expression
+                            value = f' ({value}) ' if value else value
+                            if child.get('separator') not in (' and ', ' or '):
+                                raise ValueError(
+                                    _("Wrong separator %s for %s python expression please use only ' and ' or ' or ' separator") %
+                                    (repr(child.get('separator')), repr(attribute))
+                                )
+                            separator = child.attrib.pop('separator').strip()
+
+                            if child.get('remove'):
+                                remove = child.attrib.pop('remove')
+                                if remove == value:
+                                    value = ''
+                                elif remove in value:
+                                    is_constant = remove in ('True', 'False', '1', '0')
+                                    rf_remove = rf'(( +\()*)( *{re.escape(remove)} *)((\) +)*)' if is_constant else rf'(( +\()+)( *{re.escape(remove)} *)((\) +)+)'
```

Comment:
I honestly believe that the "remove" case of way too complicated.  The proof being that you've written a naive Python parser.  I suggest the following strategy, which any developer can understand: find the first occurrence of one of the strings `f" {separator} {remove}"`, `f"{remove} {separator} "` (pick the first one that matches) and remove it from the value.  If the removal must match parentheses, then the developer must put those parentheses in `remove`.  The least amount of magic is the better, IMHO.